concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1162 @@
+# Eventing Team Code Guidelines
+
+**Document Intention** 
+
+The intention of this document is to provide style, testing and code guidelines for components of the `eventing` team. Other components in Kyma may follow totally different guidelines.
+The document is supposed to be a `living` document which can be changed by anyone in the team. However the team majority (50%) has to agree on changes for this guideline. 
+
+This guide shall be publicly available so that it can be referenced/used when reviewing pull requests. The agreements of this guide shall not affect `outside collaborators`. If a PR does not align with our guidelines, we should accept the PR as it is and apply the guidelines ourselves in a follow-up PR in order to not block external contributors. Of course, also external PRs shall meet our quality standards but we should not perform too much nit picking.
+
+For now the guidelines shall be applied to new code (PRs). There is no need to rewrite our entire code base now. This will gradually happen over time. As soon as the code reflects our guidelines, the need to look into this guide will decrease more and more.
+
+## Table of contents
+
+<!-- generate me with markdown-toc 
+```bash
+// source: https://github.com/jonschlinkert/markdown-toc
+markdown-toc -i --maxdepth 2 guidelines.md
+Do NOT TOUCH anything between the toc comments because this is used as a `marker` where to place the toc for markdown-toc.
+```
+-->
+
+<!-- toc -->
+
+- [Recommended Libraries](#recommended-libraries)
+- [Documentation Guidelines](#documentation-guidelines)
+- [Dockerfile Guidelines](#dockerfile-guidelines)
+- [Coding Guidelines](#coding-guidelines)
+- [Testing Pyramid in Kyma](#testing-pyramid-in-kyma)
+- [Testing Guidelines](#testing-guidelines)
+- [Style Guidelines](#style-guidelines)
+
+<!-- tocstop -->
+
+## Recommended Libraries
+
+The goal of this section is to describe the `desired choice` of `libraries` for testing, logging, etc.
+
+### Testing
+
+**Testing frameworks**:
+
+Use [`t.Testing`](https://pkg.go.dev/testing) for `unit` tests and for controller integration tests.
```

Comment:
in general: do we want to put the reasoning here as well?

Reply:
@k15r Do you mean for `unit` or `controller integration tests`? If for controller integration tests I would like to wait for the result of the `voting` session however (`voting options for testing framework in controller integration tests`).

But in general having a reasoning for a decision should be documented in this document.

Reply:
It also perhaps make sense to analyse which framework we already use in our code base, because it's easier to test when you already have the know-how. 
So if we mainly used one framework/technology and know how to apply it, why not keep using that ?

Reply:
@VladislavPaskar do you mean for unit or controller integration tests? And with `we` do you mean eventing  components or Kyma wide ?

For unit tests we use testing.T consistently in our evening components as far as I know. Only the matcher libraries are inconsistent (sometimes testify, sometimes gomega, sometimes pure testing.T assertions).
For controller integration tests we use ginkgo for all eventing related reconcilers.
Are you aware of any exceptions to this ?

After a longer team discussion we decided to give testing.T a try instead of ginkgo. @FriedrichWilken is working on this and is rewriting our ginkgo tests to table-driven tests using testing.T with gomega as matcher library.

Reply:
@FriedrichWilken What are the arguments ? I am interested 😃

>  But since we already have a huge testing lib with custom matchers, it would be an unreasonable amount of work to rewrite them too

True, the guideline is saying in the introduction that 
> For now, the guidelines shall be applied to new code (PRs).

So there is no need to rewrite our entire code base after merging this proposal. It shall only help us in finding a consensus.